lib/Basics/async.h

```diff
@@ -0,0 +1,132 @@
+#pragma once
+
+#include "expected.h"
+#include <coroutine>
+#include <atomic>
+#include <stdexcept>
+#include <utility>
+#include <source_location>
+
+namespace arangodb {
+
+template<typename T>
+struct async_promise;
+template<typename T>
+struct async;
+
+template<typename T>
+struct async_promise_base {
+  using promise_type = async_promise<T>;
+
+  std::suspend_never initial_suspend() noexcept { return {}; }
+  auto final_suspend() noexcept {
+    struct awaitable {
+      bool await_ready() noexcept { return false; }
+      std::coroutine_handle<> await_suspend(
+          std::coroutine_handle<> self) noexcept {
+        auto addr = _promise->_continuation.exchange(self.address());
+        if (addr == nullptr) {
+          return std::noop_coroutine();
+        } else {
+          auto continuation = std::coroutine_handle<>::from_address(addr);
+          if (addr == self.address()) {
+            self.destroy();
+            return std::noop_coroutine();
+          }
+          return continuation;
+        }
+      }
+      void await_resume() noexcept {}
+      async_promise_base* _promise;
```

Comment:
Can we make this private?

Reply:
no, this makes the code more complicated. At the same time, the user never interacts with the promise object. It is purely internal. I think we can keep it that way.